concert/tests/integration/test_pcotimestamp_addon.py

```diff
@@ -0,0 +1,160 @@
+        await self.set_random_timestamp_numbers(False)
+
+    async def _get_random_timestamp_numbers(self):
+        return self._random_numbers
+
+    async def _set_random_timestamp_numbers(self, value: bool):
+        self._random_numbers = bool(value)
+
+    async def _grab_real(self):
+        img = await super()._grab_real()
+        time = datetime.now()
+        if not await self.get_random_timestamp_numbers():
+            img = add_binary_timestamp(img, time, self.number)
+        else:
+            img = add_binary_timestamp(img, time, np.random.randint(1, 1000))
+        self.number += 1
+        return img
+
+    @transition(target='recording')
+    async def _record_real(self):
+        self.number = 1
+
+
+class TestDimaxTimestampAddon(TestCase):
+    async def asyncSetUp(self) -> None:
+        self.camera = await Camera()
+        self._data_dir = tempfile.mkdtemp()
+        self.walker = DirectoryWalker(root=self._data_dir)
+        flat_motor = await LinearMotor()
+        shutter = await Shutter()
+        self.exp = await Radiography(walker=self.walker,
+                                     camera=self.camera,
+                                     flat_motor=flat_motor,
+                                     shutter=shutter,
+                                     radio_position=0 * q.mm,
+                                     flat_position=10*q.mm,
+                                     num_projections=10,
+                                     num_flats=5,
+                                     num_darks=5)
+        self.addon = PCOTimestampCheck(self.exp)
+
+    def tearDown(self) -> None:
+        shutil.rmtree(self._data_dir)
+
+    async def test_int_to_int_array(self):
+        self.assertTrue(np.all(_int_to_int_array(123456789, 9) == [1, 2, 3, 4, 5, 6, 7, 8, 9]))
+        self.assertTrue(np.all(_int_to_int_array(1234, 9) == [0, 0, 0, 0, 0, 1, 2, 3, 4]))
+
+    async def test_add_binary_timestamp(self):
+        """
+        Tests adding timestamp to image.
+        """
+        img = np.zeros((1, 14), dtype=np.uint16)
+        time = datetime(2020, 1, 2, 3, 4, 5, 6)
+        img = add_binary_timestamp(img, time, 789)
+        timestamp = Timestamp(img)
+        self.assertEqual(timestamp.number, 789)
+        self.assertEqual(time, timestamp.time)
+
+    async def test_camera(self):
+        """
+        Tests dummy camera with timestamps.
+        """
+        await self.camera.set_random_timestamp_numbers(False)
+        async with self.camera.recording():
+            for i in range(10):
+                img = await self.camera.grab()
+                timestamp = Timestamp(img)
+                self.assertEqual(timestamp.number, i + 1)
+
+    async def run_test_addon(self):
+        """
+        Tests PCOTimestampCheck addon.
+        """
+        await self.camera.set_random_timestamp_numbers(False)
+        await self.exp.run()
+        self.assertFalse(self.addon.timestamp_incorrect)
```

Comment:
OK, and does the Acquisition finish? I am not completely sure what happens if you `break` in a consumer.

Reply:
I think so, but I will check it once more 😃

Reply:
Yes is finishes. (Otherwise the` await self.exp.run()` would also not finish).